diesel/src/sqlite/connection/bind_collector.rs

```diff
@@ -0,0 +1,119 @@
+use crate::query_builder::BindCollector;
+use crate::serialize::{IsNull, Output};
+use crate::sql_types::HasSqlType;
+use crate::sqlite::{Sqlite, SqliteType};
+use crate::QueryResult;
+
+#[derive(Debug)]
+pub struct SqliteBindCollector<'a> {
+    pub(in crate::sqlite) binds: Vec<SqliteBindValue<'a>>,
+    pub(in crate::sqlite) metadata: Vec<SqliteType>,
+}
+
+impl SqliteBindCollector<'_> {
+    pub(in crate::sqlite) fn new() -> Self {
+        Self {
+            binds: Vec::new(),
+            metadata: Vec::new(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum SqliteBindValue<'a> {
+    BorrowedString(&'a str),
+    String(Box<str>),
+    BorrowedBinary(&'a [u8]),
+    Binary(Box<[u8]>),
+    SmallInt(i16),
+    Integer(i32),
+    BigInt(i64),
+    Float(f32),
+    Double(f64),
+    Null,
+}
```

Comment:
- Binary & String are probably always constructed as String/Vec<u8>, so we could avoid reallocation.
- IIUC, some of these types don't actually exist in Sqlite, so how about removing them? (https://github.com/diesel-rs/diesel/blob/5d083cda83971ba8595c387bfb5a9998b333a3fe/diesel/src/sqlite/connection/stmt.rs#L98)
- I'd favor mapping to the rust types as they are generally clearer.

```suggestion
pub enum SqliteBindValue<'a> {
    BorrowedString(&'a str),
    String(String),
    BorrowedBinary(&'a [u8]),
    Binary(Vec<u8>),
    I32(i32),
    I64(i64),
    F64(f64),
    Null,
}
```

Reply:
Implemented in https://github.com/diesel-rs/diesel/pull/2931/commits/d29eac1ab57be3116a627a090a3f752cfdac02b8